spec/project-spec.coffee

```diff
@@ -14,6 +14,76 @@ describe "Project", ->
     atom.project.setPaths([atom.project.getDirectories()[0]?.resolve('dir')])
 
   describe "constructor", ->
+    it "enables a custom DirectoryProvider to supersede the DefaultDirectoryProvider", ->
+      remotePath = "ssh://foreign-directory:8080/"
+      class DummyDirectory
+        constructor: (@path) ->
+        getPath: -> @path
+        getFile: -> existsSync: -> false
+        getSubdirectory: -> existsSync: -> false
+        isRoot: -> true
+        off: ->
+        contains: (filePath) -> filePath.startsWith(remotePath)
+
+      directoryProvider =
+        directoryForURISync: (uri) ->
+          if uri.startsWith("ssh://")
+            new DummyDirectory(uri)
+          else
+            null
+        directoryForURI: (uri) -> throw new Error("This should not be called.")
+      atom.packages.serviceHub.provide(
+        "atom.directory-provider", "0.1.0", directoryProvider)
+
+      expect(atom.project.directoryProviders.length).toBe 2
+      expect(atom.project.directoryProviders[0]).toBe directoryProvider
```

Comment:
I would say :fire: these assertions about `.directoryProviders`, as that property is private.